TESTS/mbed_platform/FileHandle/main.cpp

```diff
@@ -0,0 +1,532 @@
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(str2_size - 1, read_ret);
+    TEST_ASSERT_EQUAL_INT(0, strncmp(str2, read_buf, str2_size - 1));
+
+    // read 3; expected read 0
+    TestFile<FS>::resetFunctionCallHistory();
+    read_ret = std::fread(read_buf, 1, str2_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_TRUE(std::feof(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(0, read_ret);
+#endif
+
+    std::fclose(file);
+}
+
+/** Test fputc and fgetc
+ *
+ *  Given already opened file
+ *
+ *  When write some data to file
+ *  Then underneath retargeting layer write function is called
+ *       fputc return written element
+ *       on failure, returns EOF and stream error is sets
+ *
+ *  When read previously written data from file
+ *  Then underneath retargeting layer read function is called
+ *       fgetc return read element
+ *       read data match previously written
+ *       on failure, returns EOF and stream error/eof is sets respectively
+ *
+ */
+void test_fputc_fgetc()
+{
+    std::FILE *file;
+    const uint32_t FS = 3;
+    TestFile<FS> fh;
+    char char_buf[3] = {'a', 'b', 'c' };
+    int ret;
+
+    file = fdopen(&fh, "w+");
+    TEST_ASSERT_NOT_NULL(file);
+#ifdef __ICCARM__
+    std::setbuf(file, NULL); // fix IAR every byte sync
+#else
+    mbed_set_unbuffered_stream(file);
+#endif
+
+    // write 1; expected written 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[0], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_EQUAL_INT(char_buf[0], ret);
+
+    // write 1; expected written 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[1], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_EQUAL_INT(char_buf[1], ret);
+
+    // write 1; expected written 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[2], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_EQUAL_INT(char_buf[2], ret);
+
+#ifndef __ICCARM__ // prevents IAR infinite loop
+    // write 1; expected written 0
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[0], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_TRUE(std::ferror(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(EOF, ret);
+#endif
+
+    fseek(file, 0, SEEK_SET);
+
+    // read 1; expected read 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fgetc(file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_EQUAL_INT(char_buf[0], ret);
+
+    // read 1; expected read 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fgetc(file);
+#ifndef __ICCARM__
+    // IAR optimize reads
```

Comment:
Is this the thing that the IAR unbuffered thing was working around, so it now behaves wrongly? Or does it break either way?

Reply:
It just behaves differently than other, but we could normalize read part and only test the first read against retargeting layer read call and remove code from ifdefs

Reply:
Is it making a single read(N) call rather than multiple read(1)? So it must be buffering? Which is what I think the original #770 was working around.

I would say that is invalid behaviour, but may not be too fatal if read(N) doesn't actually wait for N characters, but returns just what it has. The #770 fix may have been prompted by Stream::read(N) doing that wrongly.

Reply:
That kind of makes sense. Serial::read(N) waits for N characters, so Serial::scanf would have jammed. But the built-in stdin "read" always waited for 1 and returned 1, so scanf would have worked.

UARTSerial::read(N) now works correctly, returning 1-N bytes as soon as it has anything.

Maybe the set_unbuffered_stream workaround is only really needed when running against a TTY FileHandle with broken read().

Reply:
If I'm not wrong they read 512B at first 'fread' call. Can't check this now due to disabled VPN:(